// @nomicfoundation/hardhat-node-test-reporter/dist/src/formatting.js
import { styleText } from "node:util";
import { formatError } from "./error-formatting.js";
export function getInfoSymbol() {
    return styleText("blue", "\u2139");
}
export function getSuccessSymbol() {
    return styleText("green", "✔");
}
export function formatTestContext(contextStack, prefix = "", suffix = "", isSkipped = false) {
    const contextFragments = [];
    const prefixLength = prefix.length;
    for (const [i, parentTest] of contextStack.entries()) {
        const indentation = nestingToIndentationLength(parentTest.nesting);
        if (i === 0) {
            contextFragments.push(indent(prefix, indentation));
        }
        else {
            contextFragments.push("\n");
            contextFragments.push(indent("", indentation + prefixLength));
        }
        if (isSkipped && i === contextStack.length - 1) {
            contextFragments.push(styleText("cyan", `- ${parentTest.name}`));
        }
        else {
            contextFragments.push(parentTest.name);
        }
    }
    contextFragments.push(suffix);
    return contextFragments.join("");
}
export function formatTestPass(passData) {
    let msg;
    if (passData.skip === true || typeof passData.skip === "string") {
        // TODO: show skip reason
        msg = styleText("cyan", `- ${passData.name}`);
    }
    else if (passData.todo === true || typeof passData.todo === "string") {
        // TODO: show todo reason
        msg = styleText("blue", `+ ${passData.name}`);
    }
    else {
        msg = getSuccessSymbol() + styleText("gray", ` ${passData.name}`);
    }
    return indent(msg, nestingToIndentationLength(passData.nesting));
}
export function formatTestCancelledByParentFailure(failure) {
    return indent(styleText("grey", `${formatFailureIndex(failure.index)}) ${failure.testFail.name}`), nestingToIndentationLength(failure.testFail.nesting));
}
export function formatTestFailure(failure) {
    return indent(styleText("red", `${formatFailureIndex(failure.index)}) ${failure.testFail.name}`), nestingToIndentationLength(failure.testFail.nesting));
}
export function formatFailureReason(failure) {
    return `${formatTestContext(failure.contextStack, `${formatFailureIndex(failure.index)}) `, ":")}

${indent(formatError(failure.testFail.details.error), 3)}`;
}
export function formatSlowTestInfo(durationMs) {
    return ` ${styleText(["red", "italic"], `(${Math.floor(durationMs)}ms)`)}`;
}
export function formatGlobalDiagnostics(diagnostics) {
    let result = "  ";
    result +=
        styleText("green", `${diagnostics.pass} passing`) +
            styleText("gray", ` (${Math.floor(diagnostics.duration_ms)}ms)`);
    if (diagnostics.fail > 0) {
        result += "\n" + styleText("red", `  ${diagnostics.fail} failing`);
    }
    if (diagnostics.skipped > 0) {
        result += "\n" + styleText("cyan", `  ${diagnostics.skipped} skipped`);
    }
    if (diagnostics.todo > 0) {
        result += "\n" + styleText("blue", `  ${diagnostics.todo} todo`);
    }
    if (diagnostics.cancelled > 0) {
        result += "\n" + styleText("gray", `  ${diagnostics.cancelled} cancelled`);
    }
    return result;
}
function isTestOnlyMessage(message) {
    return message.includes("--test-only");
}
export function formatUnusedDiagnostics(unusedDiagnostics, testOnlyMessage) {
    const messages = unusedDiagnostics
        .map(({ message }) => {
        if (isTestOnlyMessage(message)) {
            return testOnlyMessage ?? message;
        }
        return message;
    })
        .map((message) => `  ${getInfoSymbol()} ${message}`);
    return Array.from(new Set(messages)).join("\n");
}
function formatFailureIndex(index) {
    return index.toString();
}
function nestingToIndentationLength(nesting) {
    return (nesting + 1) * 2;
}
export function indent(str, spaces) {
    const padding = " ".repeat(spaces);
    return str.replace(/^/gm, padding);
}
